library/src/layout/par.rs

```diff
@@ -1121,13 +1125,21 @@ fn line<'a>(
         let start = range.start.max(base);
         let text = &p.bidi.text[start..range.end];
         let trimmed = text.trim_end();
+        // U+200B ZERO WIDTH SPACE is used to provide a line break opportunity,
+        // we want to trim it too.
+        let trimmed = trimmed.trim_end_matches('\u{200B}');
         range.end = start + trimmed.len();
 
         // Deal with hyphens, dashes and justification.
         let shy = trimmed.ends_with('\u{ad}');
         dash = hyphen || shy || trimmed.ends_with(['-', '–', '—']);
         justify |= text.ends_with('\u{2028}');
 
+        // Deal with CJK punctuation at line ends.
+        let gb_style = is_gb_style(p.styles);
```

Comment:
queries the style chain twice for every line might be expensive. maybe we save this bool somewhere? moreover, the gb style might be set for a local text item and here you're querying the shared styles of the whole paragraph. maybe we should just add lang and region fields to `ShapedText` (although it's already getting a little big)

Reply:
Fixed in 2b3371f85f5d48edf1646c2c9a272652b29d20da

Reply:
Is it really fixed? The two new properties exist but don't seem to be used.

Reply:
Sorry I forgot that! This should be fixed now.